### PR TITLE
Added dlq and error metrics to SpannerChangestreamsReadSchemaTransformProvider

### PR DESCRIPTION
**Pull request by Dippatel98**
I have added dlq support in the Spanner Changestream read schema transform by tagging with an appropriate TupleTag. I have also added a Counter to keep track of the number of failed elements in a bundle.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Dippatel98** [comment]
r: @manavgarg, @johnjcasey, @nickuncaged1201

**Comment by Dippatel98** [comment]
Run Spotless PreCommit

**Review by manavgarg — COMMENTED** [review]
Thanks for the changes. Looks good to me overall. I was wondering if it is possible to add a unit test for this change?

**Comment by Dippatel98** [comment]
Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java [hunk 1]
```diff
@@ -193,15 +218,19 @@ public abstract static class Builder {
     }
   }
 
-  private static final class DataChangeRecordToRow extends DoFn<DataChangeRecord, Row> {
+  public static final class DataChangeRecordToRow extends DoFn<DataChangeRecord, Row> {
```
Comment (johnjcasey) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:221]: why is this changed to public?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java [hunk 2]
```diff
@@ -235,41 +264,58 @@ public void process(@DoFn.Element DataChangeRecord record, OutputReceiver<Row> r
...
+              .outputWithTimestamp(
+                  Row.withSchema(tableChangeRecordSchema)
+                      .addValue(record.getModType().toString())
+                      .addValue(record.getCommitTimestamp().toString())
+                      .addValue(Long.parseLong(record.getRecordSequence()))
+                      .addValue(rowBuilder.build())
+                      .build(),
+                  timestamp);
+        } catch (Exception e) {
+          errorsInBundle += 1;
```
Comment (johnjcasey) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:304]: Is is it possible for anything here to have an exception? I'm especially looking at .getKeysJson and .getNewValuesJson. If so, we would fail to output to the error queue

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java (hunk 1 above)
Comment (Dippatel98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:221]: In order to test it in the SpannerChangestreamsReadDlqTest. There didn't seem a straight forward way to test the dlq. So I tested the DoFn itself.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java (hunk 2 above)
Comment (Dippatel98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:304]: I discovered that it is possible for the spanner schema to change while this transform is running. It is a case which is handled in the SpannerIO successfully, but our transform doesn't handle it. That is the only mode of failure that I could come up with. As for the .getKeysJson and .getNewValuesJson they are returning the spanner row as JSON from the Mod and mapping it to a hash map. The Mod is created by the I/O itself, so I don't think there will be any formatting issues there and converting them to a hash map should be straightforward.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java (hunk 1 above)
Comment (johnjcasey) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:221]: got it. Can you annotate it as @VisibleForTesting then?

Comment (johnjcasey) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:221]: Also, you should create a test case with the scenario that artificially forces use of the DLQ, it shouldn't remain untested

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java (hunk 2 above)
Comment (johnjcasey) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:304]: Is it at all possible for there to be duplicate keys?

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java (hunk 1 above)
Comment (Dippatel98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:221]: My explanation was not clear, the dlq is tested in a way, just not the end to end workflow. I provide the DoFn with a correct and incorrect record. Then check if they are tagged properly.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java (hunk 2 above)
Comment (Dippatel98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:304]: I am not sure. Is it allowed to have two fields with the same name in spanner? If yes then there can be, other possibility is if the SpannerIO messes up.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java (hunk 1 above)
Comment (johnjcasey) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:221]: ty

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java (hunk 2 above)
Comment (johnjcasey) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:304]: tyvm. I agree its low, but poorly handle exceptions are the causes of recent outages

**Review by nickuncaged1201 — APPROVED** [review]
LGTM

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java (hunk 2 above)
Comment (Dippatel98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:307]: Done.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java (hunk 1 above)
Comment (Dippatel98) [reply sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/changestreams/SpannerChangestreamsReadSchemaTransformProvider.java:223]: Done (I have added @VisibleForTesting).

**Review by johnjcasey — APPROVED** [review]
